sherpa/models/template.py

```diff
@@ -26,149 +86,336 @@
+#
 class InterpolatingTemplateModel(ArithmeticModel):
+    """Allow parameter interpolation for a TemplateModel.
+
+    This class should not be used directly.
+
+    Parameters
+    ----------
+    name : str
+        The name of the model
+    template_model : TemplateModel instance
+        The templates to use.
+
+    See Also
+    --------
+    TemplateModel, KNNInterpolator
+
+    """
+
     def __init__(self, name, template_model):
         self.template_model = template_model
+        self.parvals = template_model.parvals
+
+        # The parameters are copied over from parvals, so we need to
+        # add them to the object, as the parent class does not do this.
+        #
         for par in template_model.pars:
             self.__dict__[par.name] = par
-            self.parvals = template_model.parvals
+
         ArithmeticModel.__init__(self, name, template_model.pars)
 
     def fold(self, data):
-        for template in self.template_model.templates:
-            template.fold(data)
+        """Ensure the templates match the data filtering.
+
+        Parameters
+        ----------
+        data : sherpa.data.Data instance
+
+        """
+        self.template_model.fold(data)
 
     @modelCacher1d
     def calc(self, p, x0, x1=None, *args, **kwargs):
         interpolated_template = self.interpolate(p, x0)
         return interpolated_template(x0, x1, *args, **kwargs)
 
+    def interpolate(self, point, x_out):
+        """Calculate the model combination for the given parameter values.
+
+        If the given point matches a template than that is used, otherwise
+        weight the templates in some manner.
+
+        Parameters
+        ----------
+        point : sequence of float
+            The parameter values to use. This must match the size and
+            ordering of the parameters of the model.
+        x_out : sequence of float
+            The x values to use for the model evaluation.
+
+        Returns
+        -------
+        model : TableModel instance
+            The model to use.
+
+        """
+
+        raise NotImplementedError
+
 
 class KNNInterpolator(InterpolatingTemplateModel):
+    """Use K-nearest neighbors interpolation for parameters.
+
+    Select the template to use based on the K-nearest neighbors
+    algorithm
+    https://en.wikipedia.org/wiki/K-nearest_neighbors_algorithm
+
+    Parameters
+    ----------
+    name : str
+        The name of the model
+    template_model : TemplateModel instance
+        The templates to use.
+    k : int or None, optional
+        The k to use.
+    order : int, optional
+        The order to use.
+
+    See Also
+    --------
+    TemplateModel
+
+    """
     def __init__(self, name, template_model, k=None, order=2):
-        self._distances = {}
         if k is None:
-            self.k = 2*template_model.parvals[0].size
+            # TODO: what happens if no params? What is .size meant to be?
+            self.k = 2 * template_model.parvals[0].size
```

Comment:
The TODO can be removed, because this is safe. `parvals` is of shape (nelem, npars) and `template_model.parvals[0].size` is the same as `template_model.parvals.shape[1]` (but I find the latter much easier to read). So, if npars==0, then self.k will be 0, which should be safe and the default is to use 2 * the number of "dimentions" (i.e. npars), which makes sense.

Reply:
This note was added when I was trying to understand the code. I'll change it to your suggestion.